Integrate vllm and inference engine (neural speed)

jiafuzha opened the pull request.

jiafuzha commented:
@KepingYan @carsonwang  @xwu99 please help review.

xwu99 commented:
> @KepingYan @carsonwang @xwu99 please help review.

Great work! I will check this week and let you know.

xwu99 commented on vllm-ext/vllm/extension/ns/__init__.py:
```diff
@@ -0,0 +1,174 @@
+
+
+def _verify_quntization(self):
+    if self.quantization is not None and self.quantization == "ns":
```
```suggestion
    if self.quantization == "ns":
```

jiafuzha replied:
good catch

xwu99 commented on vllm-ext/inference_engine/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,134 @@
+option(IE_AVX                    "inference_engine: enable AVX"                                     ON)
+option(IE_AVX2                   "inference_engine: enable AVX2"                                    ON)
+option(IE_F16C                   "inference_engine: enable F16C"                                    ON)
+option(IE_AVX512                 "inference_engine: enable AVX512"                                  OFF)
```
Do we build AVX512 by default? I am not sure the usage of the option here.

jiafuzha replied:
It's actually not used for now. Instrinsics are checked dynamically during run-time. Will keep it here for later.

xwu99 commented on vllm-ext/vllm/extension/ns/__init__.py:
```diff
@@ -0,0 +1,174 @@
+model_loader = importlib.import_module("vllm.model_executor.model_loader")
+importlib.reload(model_loader)
+
+logger.info("__ns extension: use ns model loader for ns model, %s", NSModelLoaderV2.__name__)
```
could you summerize a list of which properties are monkey-patched with NS classes in the PR descriptions ?

jiafuzha replied:
All monkey-patches are in this __init__.py. They are,

__init__.py:52] __ns extension: add ns quantization config, NSQuantConfig
__init__.py:105] __ns extension: use ns model loader for ns model, NSModelLoaderV2
__init__.py:116] __ns extension: replace LlamaModel with ns LLamaModel, NSLLamaModel
__init__.py:136] __ns extension: use ns cache engine for ns, NSCPUCacheEngine
__init__.py:146] __ns extension: replace execute_model in cpu_model_runner, execute_model
__init__.py:171] __ns extension: replace BlockSpaceManager.get_block_space_manager_class in vllm.core.interfaces with get_block_space_manager_class

xwu99 commented on llm_on_ray/inference/predictors/vllm_predictor.py:
```diff
@@ -43,15 +49,44 @@ def __init__(self, infer_conf: InferenceConfig, max_num_seqs):
+        if infer_conf.vllm.extension == "ns":
+            logger.warn("applying neural speed extension to vllm ...")
+            try:
+                from vllm.extension import ns as ns
```
```suggestion
                from vllm.extension import ns

jiafuzha replied:
good catch

xwu99 commented on vllm-ext/vllm/extension/ns/model/ns_model.py:
```diff
@@ -0,0 +1,324 @@
+            physical_cores = psutil.cpu_count(logical=False)
+            # reserve one core for non-ns tasks
+            physical_cores = physical_cores if physical_cores <= 1 else physical_cores - 1
+            threads = int(os.environ.get(_NS_NUM_THREADS, str(physical_cores)))
```
consider set the cores for ns according to `cpus_per_worker` in the inference config, all other predictors use this config to assign cpu resources.

jiafuzha replied:
It's NS threads which is different from ray cpus. I want to separate them.

jiafuzha replied:
Just came to my mind that bigger --num-cpus value causes ray starts up more background processes, like client.poll or server.poll processes which hurts overall performance. That's the reason why I set num-cpus to 1 to reduce them. Need to find an elegant way to balance them.

xwu99 commented on vllm-ext/vllm/extension/ns/model/ns_model.py:
```diff
@@ -0,0 +1,324 @@
+        # get available cores
+        try:
+            max_prompt_tokens = int(os.environ.get(_NS_MAX_PROMPT_TOKENS, "8192"))
+            # cpus_per_work is set to 1 for better ns perf so it's inappropriate to use ray to get available cores
```
you can directly use `cpus_per_work` to set cpu cores for ns, no need to set cpus_per_work to 1 and use another env for ns.

jiafuzha replied:
no, we cannot. cpus_per_worker value is set to OMP_NUM_THREADS. I actually need to set OMP_NUM_THREADS to 1 to get better perf.

xwu99 commented:
`vllm-ext/inference_engine/cpp/models/application` this directory should be in `vllm-ext/inference_engine/cpp` according to neural_speed code structure

jiafuzha commented:
> `vllm-ext/inference_engine/cpp/models/application` this directory should be in `vllm-ext/inference_engine/cpp` according to neural_speed code structure

I am not 100% follow neural_speed structure since we only take it as inference engine.

jiafuzha commented:
It's code before merge. I've addressed all review comments in after merge branch. So, let me close this one and submit a new PR in after merge branch.
